Improve Revs Performance by Avoiding Repeated Tree Checkouts and Adding Version Indexing

suratkhan opened the pull request:
This PR solves #1559 by refactoring the `Revs` implementation to improve performance by:

- Pre-building an index of `Rev` structs from commit trees.
- Eliminating repeated `checkout_tree` calls.
- Using binary search for efficient `find_version` lookups.
- Supporting iteration over a cached, sorted list of versions.

suratkhan commented:
@smoelius, here’s a summary of the changes I made in this PR:

- Refactored `Revs` to pre-index all `Rev` structs at initialization, reducing repeated repository traversal.
- Replaced `checkout_tree` logic with direct blob access and parsing for `Cargo.toml` and `rust-toolchain`.
- Improved version lookup speed by introducing binary search in `find_version`.
- Implemented a proper `RevIter` for forward iteration over all versions.
- Maintained existing functionality, output format, and test coverage.

smoelius reviewed (commented):
I left some comments below. I think they may require significant changes, so I'm going to pause here and let you reflect on them.

I really appreciate you working on this.

suratkhan commented:
hello @smoelius, I have applied the things you mentioned in the comments. Here is a summary of what I did:

1.  Refactored `revs.rs`:
    *   Replaced the full commit history scan and upfront version indexing with a "lazy-loading" approach. Commits are loaded in batches as needed.
    *   Implemented binary search over commits (instead of versions) to efficiently find the revision corresponding to a specific `--rust-version`.
    *   Added a dedicated `find_latest_rev` function that scans backwards from HEAD for the latest version (used when `--rust-version` is omitted).
    *   Removed the old `build_version_index` logic and the `RevIter` struct.
2.  Updated `mod.rs`:
    *   Modified `upgrade_package` to call the new `revs.find_version()` and `revs.find_latest_rev()` methods.

smoelius commented on dylint/src/package_options/revs.rs:
```diff
@@ -127,58 +291,135 @@ mod test {
+        let result = revs.find_version(ancient_version).unwrap();
+        // Depending on history, this might find the oldest known version or None if history starts
+        // later. If it finds *a* version, it should be the oldest one available.
+        if let Some(rev) = result {
```
I like the idea of having a test that finds the earliest commit `find_version` possibly can!

But this check is kind of weird.

From what I can tell, the earliest commit in the `rust-clippy` repository is this one: https://github.com/rust-lang/rust-clippy/commit/37226273a7a5b2119daaab06d253f93b6813b881

This appears to be the earliest commit with a `rust-toolchain` file: https://github.com/rust-lang/rust-clippy/commit/40151d91af6539087383038e0cc95ea7ce3b9b91

And this appears to be the earliest commit with a `rust-toolchain` file with a modern format, e.g., a channel: https://github.com/rust-lang/rust-clippy/commit/f03edfd7a1a02c696a600a1776aa847edd98c378

Would it be possible to have the test find one of those commits, or some commit close thereto, and handle it appropriately?

smoelius reviewed (commented):
You've put a lot of work into this! Thank you for that.

I hope you can forgive me, I'm going to have to look at this a few times, just because of its complexity.

Thanks a lot for working on this!

suratkhan commented:
Hi @smoelius, good to see you again! I hope you're doing well.

Thanks for the detailed review! I worked on this about two weeks ago, so I'll need to take another look to refresh my memory on what I did and why. Hopefully, I remember everything. I'll try my best to address the feedback and get things done within this week.

In the meantime, here (#1606) is the PR that fixes the error messages.

suratkhan commented:
Hi @smoelius,

Sorry I’ve been inactive for the last couple of weeks—I was on vacation. I’ve just started back this week and made the changes you requested. After rebasing onto master and committing, I’m now seeing some new CI failures. Could you take a look and let me know whether they’re related to my changes, and if so how I should address them?

Thanks!

smoelius commented:
> I’m now seeing some new CI failures. Could you take a look and let me know whether they’re related to my changes, and if so how I should address them?

Sorry, @suratkhan. The answer is not immediately obvious to me. I will look at this more closely later on today.

smoelius commented on dylint/src/package_options/revs.rs:
```diff
@@ -1,123 +1,276 @@
+            } else {
+                return Ok(None);
+            }
+        };
```
I think the problem is here. It looks like this code is reading the entire `rust-toolchain` file and treating it as the toolchain channel. I think `rust-toolchain` files used to work that way, but no longer. I recommend using the `toolchain_channel` function to get the toolchain channel.

suratkhan commented:
Hi @smoelius,

I tried to address the issue as you mentioned, but unfortunately it didn’t work out. I'm not very familiar with the usage of the `toolchain_channel` function. When you have some time, could you please provide a bit more detail or guidance on how to use it?

Thank you!

smoelius commented:
Sorry for not being more specific. This is the function I was referring to: https://github.com/trailofbits/dylint/blob/56e2dcbf22b73e9b42d658b15ac5318bf7e4be37/internal/src/clippy_utils.rs#L68

You call it with the path of a `rust-toolchain` or `rust-toolchain.toml` file, and it extracts the `channel` field.

Here are two examples where it is used:
- https://github.com/trailofbits/dylint/blob/6852355c8d73cbd20cb91abba9e45175d6ba722d/dylint/src/package_options/revs.rs#L49
- https://github.com/trailofbits/dylint/blob/6852355c8d73cbd20cb91abba9e45175d6ba722d/dylint/src/package_options/revs.rs#L101
